### ISSUE TITLE
RFE request : choosing which shell to invoke

### ISSUE THREAD

**Issue by rwthompsonii** [eranif/codelite#1097]
i have a build system that I'd like to use with codelite, but I'm running into issues because a ton of the makefile and infrastructure that I'm using with the custom settings depends on #!/bin/bash.  However, codelite is pretty insistent on using /bin/sh.  I'm currently aliasing /bin/sh to /bin/bash to force it to the behavior I want, but to just ask would be kind of nice on initial configuration.

**Comment by eranif** [comment]
Where do you see this?
Please provide a full list of where you think CodeLite is insisting on `#!/bin/sh`

**Comment by rwthompsonii** [comment]
Wow.  You're fast.  I just set up my new CodeLite install, so I may be messing it up.  Right now I have a sourceme script setup in my top level project directory that I can run source SOURCEME.SH on and it will setup my makefiles' required environment variables.  The issue is that when I attempt to source it using the pre-build menu, it tells me that "source: command not found", which I'm assuming is because the pre-build step is run as a /bin/sh.  I try to work around it with #!/bin/bash at the top (no effect), executing bash directly as one of the steps (hangs), finally just hardcode the exports, but those env variables aren't available to the build step(?).  I just closed codelite and sourced the shell script first and then rerun codelite, and voila, variables are there.

**Comment by eranif** [comment]
Yes, this is how I usually works (at work...)
our build environment also depends on environment variables, so I run `. Environment` from a terminal and then launches CodeLite from that terminal and it works

**Comment by rwthompsonii** [comment]
It's compiling now, which is already a massive improvement over Eclipse, so I'm happy. Just a thought if you had free time is the ability to choose somewhere in settings "Default Shell" with a text box for a path to the executable. Not sure how painful that would be on your end, if stuff is coded directly with .sh that might suck to fix.

**Comment by eranif** [comment]
Usually, CodeLite runs external tools like this:
`/bin/sh -c ....` is this what you mean?

**Comment by rwthompsonii** [comment]
Or, another idea, is to provide a path to a script already on disk to run before the build. That'd be hot. Especially if it somehow keeps my environment variables for the build step. But since it's not preventing me from working, no bigs.

**Comment by rwthompsonii** [comment]
Yes. Instead of /bin/sh, reference the provided executable (maybe with exec?) so I can override with bash or another shell.

**Comment by eranif** [comment]
I think that ideally, workspace->workspace settings->environment variables is the best location to enable this (and also, provide a way to override this on the project level)

Probably parsing the script and generate a CodeLite syntax from it
Can you paste here the SOURCEME.SH script that you are using (to see a real world use case)

**Comment by rwthompsonii** [comment]
yeah give me a minute to obfuscate some of it (I wish I could work on open source stuff)

**Comment by rwthompsonii** [comment]
the script that runs accepts two arguments, one for switching between development builds and release builds, and the other for switching the architecture to feed into the compiler (I work on embedded systems), such as mips or x86 or arm.  

it's usage message portion is something like this: 
# Usage message
PROGRAM=`basename $BASH_SOURCE`
Usage () {
    echo "Usage: $PROGRAM [architecture] --[development|release]"
    echo "--development            don't strip out logging"
    echo "architecture  sets up the right environment variables and directories for obj files"
}

another section which I don't think would work under /bin/sh (i'm not a shell expert):

# Make sure build/stuff/bin is in the path
if [ ! `echo $PATH | grep "$build/stuff/bin"` ]; then
    export PATH=<put it in the path and then execute some other crap>
fi

the file has at least 1000 lines of doing stuff like checking OS (we develop on everything but Windows), setting up intermediate build directories and whatnot, all with what I believe to be bash syntax. 

Also, jesus I had to look at that three times to make sure it was safe to release to the wild.  There's a lot of shit in that file.

If you need more details I may be able to email them directly but I probably shouldn't post them here.  In particular, I'm not that familiar with the differences between bash and sh (aliased to dash on my workstation).  If you are, I could definitely look for other examples.

**Comment by eranif** [comment]
Thanks for the info, its more than enough.
So your script is a complex one, and not just a bunch of `export SOME=THING` so writing a parser to parse it, is not an option here.

However, using the following method I think that CodeLite can import it like this:

1. Open a terminal 
2. Take an environment variables snapshot
3. Run the user script
4. Take second environment variables snapshot
5. Diff the results

Ofc, steps 1-5 will be done automatically by CodeLite using a nice UI wrapping it

**Comment by rwthompsonii** [comment]
Indeed, that sounds pretty good to me.  

Thanks for being so responsive, it's nice.

**Comment by zozigizo** [comment]
I'm also having problem with this - I think the reason because for makefile CodeLite runs external tools like this:
`/bin/sh -c make ` with no way to configure it to `/bin/bash make`
more specifically I'm trying to compile arduino project with [https://github.com/sudar/Arduino-Makefile](url).

the result is from command line everything work perfect. from Codelite a lot of problems I think it's because bash vs sh?

A lot of people want arduino support: [https://github.com/eranif/codelite/issues/537](url)

I believe by adding bash instead of sh support for makefile calling it may work? I've almost got it working... so close...
